### PR TITLE
[1/n] Add LightningFetcher

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduce the LightningFetcher.

Next PR will add fault tolerant training components.

Parts of https://github.com/PyTorchLightning/pytorch-lightning/issues/8867

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/fetching.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+from typing import Any, Generator, List, Optional, Tuple
+
+from torch.utils.data.dataloader import DataLoader
+
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class AbstractFetcher(ABC):
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:26]:
```suggestion
class AbstractDataFetcher(ABC):
```

Can we call it data-fetcher everywhere to make sure we always know what it fetches?

File: pytorch_lightning/utilities/fetching.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    This class is used to control batch fetching flow.
+    """
+
+    @abstractmethod
+    def fetching_function(self) -> Generator:
+        pass
+
+    def __init__(
+        self,
+        prefetch_batches: int = 1,
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:38]: does `prefetch_batches=1` mean that there is one additional batch pre-fetched (while the current one is processed) or is this the current behaviour with sequential fetching? If it is the current, I think this is ambigious and it should rather be 0 for this behaviour.

I would strongly advocate for defaulting to the current behaviour and making this opt-in always since there are many scenarios, where only a single batch fits into memory.

File: pytorch_lightning/utilities/fetching.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        return self.batches.pop(0)
+
+    @property
+    def loaders(self) -> List[DataLoader]:
+        if not self._has_setup:
+            raise MisconfigurationException("The Fetcher should be setup with a ``dataloader``.")
+        if isinstance(self.dataloader, CombinedLoader):
+            loaders = self.dataloader.loaders
+        else:
+            loaders = [self.dataloader]
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:67]:
```suggestion
        elif isinstance(self.dataloader, (tuple, list)):
            loaders = self.dataloader
        else:
            loaders = [self.dataloader]
```
to also support sequential loaders as in validation, not sure if this is necessary.

File: pytorch_lightning/utilities/fetching.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        return loaders
+
+    @property
+    def loader_iters(self) -> List[Iterator]:
+        if not self._has_setup:
+            raise MisconfigurationException("The Fetcher should be setup with a ``dataloader``.")
+        if isinstance(self.dataloader, CombinedLoader):
+            loader_iters = self.dataloader_iter.loader_iters
+        else:
+            loader_iters = [self.dataloader_iter]
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:77]: It the above was accepted for sequential loaders, this has to be applied here as well :)

Comment (justusschock) [reply pytorch_lightning/utilities/fetching.py:77]: Accessing `self.dataloader_iter` is invalid outside of `__iter__`

File: pytorch_lightning/utilities/fetching.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        def collect_state(iterator: Iterator):
+            return iterator.state
+
+        return apply_to_collection(self.loader_iters, Iterator, collect_state)
+
+    def __iter__(self) -> Generator[Tuple[Any, bool], None, None]:
+        if self.dataloader is None:
+            raise MisconfigurationException("The iterate hasn't been provided. HINT: Did you call setup function ?.")
+        self.reset()
+        self.dataloader_iter = iter(self.dataloader)
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:91]: If storing this into a variable, rather already define that variable in `__init__`

File: pytorch_lightning/utilities/fetching.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        return self.fetching_function()
+
+    def reset(self) -> None:
+        self.batches: List = []
+        self.dataloader: Optional[Iterable]
+        self.fetched: int = 0
+        self.done: bool = False
+        self.has_raised: bool = False
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:99]: Either define those in init, or at least add typing stubs at the beginning of the class (above init)

File: pytorch_lightning/utilities/fetching.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:26]: good point !

File: pytorch_lightning/utilities/fetching.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:38]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:77]: Done ! Yes, I will have to add support for Collection, but this is still internal use.

File: pytorch_lightning/utilities/fetching.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:91]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:99]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:67]: Done ! This is not necessary for now. At least until this become part of the public API.

File: pytorch_lightning/utilities/fetching.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:38]: @justusschock I made the update, but I am not sure to like it. I mean this isn't final yet.
@awaelchli @ananthsub @carmocca Any thoughts ?

**Review by justusschock — APPROVED** [review]

**Comment by carmocca** [comment]
Is there some bigger picture somewhere of how this is going to be used, what will this replace, and how will users interact with or override it?

**Comment by tchaton** [comment]
> Is there some bigger picture somewhere of how this is going to be used, what will this replace, and how will users interact with or override it?

Hey @carmocca, yes definitely.

The LightningFetcher was introduced through this RFCs: https://github.com/PyTorchLightning/pytorch-lightning/issues/8867 and will be used to replace the `prefetch_function`. This should also be used for validation purposes.

There is a long way before we make this object public API, however it is there to solve 3 primary issues:

* Handle Fault Tolerant dataloader dictionary state dict aggregation as used within this PoC:  https://github.com/PyTorchLightning/pytorch-lightning/pull/8476/files and this PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/8891
* Add support for inter batch parallelism 
* Provide a way for users to customise data being passed to their training_step.

Best,
T.C

File: pytorch_lightning/utilities/fetching.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        self.reset()
+
+    def setup(self, dataloader: DataLoader, **kwargs) -> None:
+        if not isinstance(dataloader, (DataLoader, CombinedLoader)):
+            raise MisconfigurationException(
+                "The `DataFetcher` should be setup with an instance of a PyTorch ``DataLoader``."
+            )
+        self.dataloader = dataloader
+
+    def add_batch(self, batch) -> None:
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/fetching.py:62]:
```suggestion
    def add_batch(self, batch: Any) -> None:
```

File: pytorch_lightning/utilities/fetching.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        self.batches.append(batch)
+
+    def fetch_batch(self) -> Any:
+        return self.batches.pop(0)
+
+    @property
+    def loaders(self) -> List[DataLoader]:
+        if not self.dataloader:
+            raise MisconfigurationException(
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/fetching.py:71]: Add a unit test to check this raise?

File: tests/utilities/test_fetching.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import pytest
+from torch import tensor
+from torch.utils.data import DataLoader, IterableDataset
+
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities.fetching import LightningFetcher
+
+
+@pytest.mark.parametrize("use_combined_loader", [False, True])
+def test_prefetch_iterator(use_combined_loader):
```
Comment (SeanNaren) [inline tests/utilities/test_fetching.py:23]: How will the LightningFetcher be sent to the Trainer? From offline discussions I recall it being like so:

```python
trainer = Trainer(lightning_fetcher=CustomLightningFetcher())
```
is that so? If so cool :)

File: pytorch_lightning/utilities/fetching.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:62]: Done !

File: tests/utilities/test_fetching.py (hunk 9 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:23]: Yes, something like that in the future :)

Comment (tchaton) [reply tests/utilities/test_fetching.py:23]: trainer = Trainer(fetched_batches=[0, 1, 2, "auto", CustomLightningFetcher(batch_processor=...)])

**Review by SeanNaren — APPROVED** [review]
Awesome!

File: pytorch_lightning/utilities/fetching.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    @abstractmethod
+    def fetching_function(self) -> Generator:
+        pass
+
+    def __init__(
+        self,
+        prefetch_batches: int = 1,
+    ) -> None:
+        if not isinstance(prefetch_batches, int) or (isinstance(prefetch_batches, int) and prefetch_batches < 0):
+            raise MisconfigurationException("`prefetch_batches` should at least be 1.")
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/fetching.py:41]: The error message here says the argument should be at least 1, but the if triggers the error only if the argument is less than zero. So it can also be zero? Or the if should be `prefetch_batchs <= 0`?

File: pytorch_lightning/utilities/fetching.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    This class is used to control batch fetching flow.
+    """
+
+    @abstractmethod
+    def fetching_function(self) -> Generator:
+        pass
+
+    def __init__(
+        self,
+        prefetch_batches: int = 1,
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/fetching.py:38]: If the default is one then `self.prefetch_batches` becomes 2, is that correct?

File: pytorch_lightning/utilities/fetching.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:41]: good catch !

File: pytorch_lightning/utilities/fetching.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: bad merged, sorry. Should be 0 by default.

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by ananyahjha93 — APPROVED** [review]
